napi-inl.h

```diff
@@ -3705,6 +3705,28 @@ inline const napi_node_version* VersionManagement::GetNodeVersion(Env env) {
   return result;
 }
 
+#if (NAPI_VERSION > 2)
+////////////////////////////////////////////////////////////////////////////////
+// Cleanup class
+////////////////////////////////////////////////////////////////////////////////
+
+inline CleanupHook::CleanupHook(Env env) : _env(env) {
+  napi_status status = napi_add_env_cleanup_hook(_env, NapiHook, static_cast<void*>(this));
+  NAPI_THROW_IF_FAILED_VOID(_env, status);
+}
+
+inline CleanupHook::~CleanupHook() {
+  napi_status status = napi_remove_env_cleanup_hook(_env, NapiHook, static_cast<void*>(this));
+  NAPI_THROW_IF_FAILED_VOID(_env, status);
+}
+
+inline void CleanupHook::NapiHook(void* arg) {
+  CleanupHook* self = static_cast<CleanupHook*>(arg);
+  self->Cleanup();
+  delete self;
```

Comment:
This is unsafe. If the CleanupHooks is created on non-heap, this statement will cause a crash.

Reply:
As far as I know, hooks added by `napi_remove_env_cleanup_hook` API are called after the javascript code is done, so even if a subclass is created in the non-heap area such as stack, finally the destructor would be called so those hooks are removed. But honestly, I'm not sure 100% of the life cycle of the node so correct me if there's any wrong. :)

Reply:
You could only allow on-heap object by using private constructor with static factory method.

Reply:
And `final`. :) I don’t understand why we’d want a heap-object-only restriction, though?

Reply:
> And final. :) I don’t understand why we’d want a heap-object-only restriction, though?

Am I missing something? I'm not sure if but this line always calls `delete self` whenever N-API Env hooks callback is called. Then, it will cause a crash. (e.g. if the object is created on static area) Therefore, we should remove this line and rewrite another way at least(if we don't want a heap-object-only).